File: scala/133.scala
object Solution {
    def cloneGraph(graph: Node): Node = {
        val nodes = Array.ofDim[Node](101)
        def dfs(g: Node): Node = {
          if (nodes(g.value-1)==null) {
            nodes(g.value-1) = new Node(g.value)
            nodes(g.value-1)
          } else return nodes(g.value-1)
          nodes(g.value-1).neighbors = g.neighbors.map(n => dfs(n)).toList
          nodes(g.value-1)
        }
        if (graph == null) return null
        dfs(graph)
        nodes.head
    }
}

File: leetcodeJava/133.java
package leetcodeJava;

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        var visited = new Node[101];
        return dfs(node, visited);
    }
    
    private Node dfs(Node node, Node[] visited) {
        if (visited[node.val] != null) {
            return visited[node.val];
        }
        var n = new Node(node.val);
        visited[n.val] = n;
        for (var neighbor : node.neighbors) {
            n.neighbors.add(dfs(neighbor, visited));
        }
        return n;
    }
}
